src/Agent.Worker/Build/BuildJobExtension.cs

```diff
@@ -339,6 +342,60 @@ private string ConvertToLegacyRepositoryType(string pipelineRepositoryType)
             }
         }
 
+        private string GetDefaultRepoLocalPathValue(IExecutionContext executionContext, IList<Pipelines.JobStep> steps, TrackingConfig trackingConfig, RepositoryInfo repoInfo)
+        {
+            string selfRepoPath = null;
+            // For saving backward compatibility with the behavior of the Build.RepoLocalPath that was before this PR https://github.com/microsoft/azure-pipelines-agent/pull/3237
+            // We need to change how we set the default value of this variable
+            // We need to allow the setting of paths from RepositoryTrackingInfo for checkout tasks where path input was provided by the user
+            // and this input is not point to the default location for this repository
+            // This is the only case where the value of Build.RepoLocalPath variable is not pointing to the root of sources directory /s.
+            // The new logic is not affecting single checkout jobs and jobs with multiple checkouts and default paths for Self repository
+            if (RepositoryUtil.HasMultipleCheckouts(executionContext.JobSettings))
+            {
+                // get checkout task for self repo
+                var selfCheckoutTask = GetSelfCheckoutTask(steps);
+
+                // Check if the task has path input with custom path, if so we need to set as a value of selfRepoPath the value of SourcesDirectory from RepositoryTrackingInfo
+                if (IsCheckoutToDefaultPath(trackingConfig, repoInfo, selfCheckoutTask))
+                {
+                    {
+                        selfRepoPath = trackingConfig.RepositoryTrackingInfo
+                            .Where(repo => RepositoryUtil.IsPrimaryRepositoryName(repo.Identifier))
+                            .Select(props => props.SourcesDirectory).FirstOrDefault(); 
+                    }
+                }
+            }
+            // For single checkout jobs and multicheckout jobs with default paths set selfRepoPath to the default sources directory
+            if (selfRepoPath == null)
+            {
+                selfRepoPath = trackingConfig.SourcesDirectory;
+            }
+
+            return selfRepoPath;
+        }
+
+        private bool IsCheckoutToDefaultPath(TrackingConfig trackingConfig, RepositoryInfo repoInfo, TaskStep selfCheckoutTask)
+        {
+            string path;
+            string selfRepoName = RepositoryUtil.GetCloneDirectory(repoInfo.PrimaryRepository.Properties.Get<string>(Pipelines.RepositoryPropertyNames.Name));
+            string defaultRepoCheckoutPath = Path.GetFullPath(Path.Combine(trackingConfig.SourcesDirectory, selfRepoName));
+
+            return selfCheckoutTask != null
+                    && selfCheckoutTask.Inputs.TryGetValue(PipelineConstants.CheckoutTaskInputs.Path, out path)
+                    && !string.Equals(Path.GetFullPath(Path.Combine(trackingConfig.BuildDirectory, path)),
```

Comment:
Is 'IsCheckoutToDefaultPath' a proper name if there is a  'path != defaultRepoCheckoutPath' check?

Reply:
Oh, yeah not a very good name for the method